Add universal binary support for macOS

startergo opened the pull request:
- Modified configure script to use Command Line Tools SDK
- Added -arch x86_64 -arch arm64 flags for universal binaries
- Updated Makefile to support both Intel and Apple Silicon architectures
- Fixed conditional compilation for ARM64 vs x86 architectures

startergo committed:
Add universal binary support for macOS

- Modified configure script to use Command Line Tools SDK
- Added -arch x86_64 -arch arm64 flags for universal binaries
- Updated Makefile to support both Intel and Apple Silicon architectures
- Fixed conditional compilation for ARM64 vs x86 architectures

joevt commented on lib/configure:
```diff
@@ -137,11 +137,13 @@ case $sys in
 		echo >>$c '#define PCI_HAVE_64BIT_ADDRESS'
-		echo_n " i386-ports"
-		echo >>$c '#define PCI_HAVE_PM_INTEL_CONF'
+		if [ "$cpu" = "i386" -o "$cpu" = "x86_64" ] ; then
+			echo_n " i386-ports"
```
I'm not sure this change is necessary or correct.

PowerPC architecture can use `intel-conf1` access method defined by `i386-ports` if running under Rhapsody in Tiger, Leopard, and Snow Leopard because pciutils uses the DirectHW kext which runs in the kernel as Intel architecture.

The `intel-conf1` access method will output a warning when it can't be used so there's no reason to exclude it.

Note that the rest of the configure file likes to use `case` instead of `if` when testing the cpu.

startergo replied:
PowerPC systems can use intel-conf1 access method under certain conditions
the configure script uses `case $cpu in i?86|x86_64)` in multiple places, all outputting " i386-ports" 
The intel-conf1 access method works for both i386 and x86_64 architectures, and the naming reflects the method type rather than the CPU architecture specifically. so you are correct.

joevt commented on lib/configure:
```diff
@@ -137,11 +137,13 @@ case $sys in
 		LIBRESOLV=
 		LIBEXT=dylib
-		SYSINCLUDE=$(xcrun --sdk macosx --show-sdk-path 2> /dev/null || : )/usr/include
+		SYSINCLUDE=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
```
Does `xcrun` not work with CommandLineTools?
Consider the Copilot suggestion.

startergo replied:
done

joevt commented on Makefile:
```diff
@@ -2,7 +2,8 @@
 
 OPT=-O2
-CFLAGS=$(OPT) -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes
+CFLAGS=$(OPT) -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -arch x86_64 -arch arm64 -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes
+TARGET_ARCH=-isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -arch x86_64 -arch arm64
```
Why are `CFLAGS` and `TARGET_ARCH` the same?

I think these Makefiles need to still work for i386 and ppc and ppc64 architectures.
Maybe we need to check SDK versions to see which architectures can be built?
One method Makefile files or Configure files use to do that is to run a test compile/link command and check for error.

startergo replied:
done

startergo committed:
Improve SDK detection robustness in configure script

- Add xcrun fallback for SDK path detection on macOS
- Use xcrun --show-sdk-path as primary method with Command Line Tools fallback
- This ensures compatibility across different macOS development environments

startergo committed:
Improve macOS Makefile architecture detection

- Separate CFLAGS from TARGET_ARCH properly to avoid redundancy
- Add intelligent architecture detection that tests both compilation and linking
- Support fallback to native architecture when universal binaries aren't fully supported
- Use Command Line Tools SDK with Xcode SDK fallback
- Prevent linking errors by only including architectures with complete library support
- Maintain compatibility with older macOS systems while supporting modern architectures

joevt commented:
I'll do some test builds tomorrow on Intel 10.5 and up and PPC 10.4, 10.5, 10.6. Have you done any Apple Silicon builds? Does pciutils work on Apple Silicon? If not, then that will be a different commit/pull request, if it's fixable.

startergo commented:
> I'll do some test builds tomorrow on Intel 10.5 and up and PPC 10.4, 10.5, 10.6. Have you done any Apple Silicon builds? Does pciutils work on Apple Silicon? If not, then that will be a different commit/pull request, if it's fixable.

Lspci comes with no accesible methothd found. Perhaps apart from sip I need debug=0x144?

joevt commented:
> Lspci comes with no accesible methothd found. Perhaps apart from sip I need debug=0x144?

DirectHW is installed? Used sudo? What does `sudo lspci -A detect` output?

Intel Mac:
```
sudo lspci -A detect
Trying method darwin......using AppleACPIPlatformExpert...OK
Trying method darwin2......using IOPCIBridge...OK
Trying method darwin3......using DirectHW IOPCIBridge...OK
Trying method intel-conf1......sanity check...outside the Asylum at 0/00/0...OK
Trying method intel-conf2......No.
```
